lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+    }
+}
+
+// Important: when converting String to CGOString, the caller is responsible for manually freeing
+// the memory at CGOString.data.
+impl From<String> for CGOString {
+    fn from(mut s: String) -> CGOString {
+        let cs = CGOString {
+            data: s.as_mut_ptr(),
+            len: s.len() as u16,
+        };
+        // Tell Rust to forget about the underlying memory, so that the caller can free it later.
+        std::mem::forget(s);
+        cs
+    }
+}
+
+// connect_rdp establishes an RDP connection to go_addr with the provided credentials and screen
+// size. If succeeded, the client is internally registered under client_ref. When done with the
+// connection, the caller must call close_rdp.
+#[no_mangle]
+pub extern "C" fn connect_rdp(
+    go_addr: CGOString,
+    go_username: CGOString,
+    go_password: CGOString,
+    screen_width: u16,
+    screen_height: u16,
+    client_ref: i64,
+) -> CGOError {
+    // Convert from C to Rust types.
+    let addr = String::from(go_addr);
+    let username = String::from(go_username);
+    let password = String::from(go_password);
+
+    // Connect and authenticate.
+    let tcp = match TcpStream::connect(&addr) {
+        Ok(tcp) => tcp,
+        Err(e) => return format!("failed TCP connection to {}: {:?}", &addr, e).into(),
+    };
+    let tcp_fd = tcp.as_raw_fd() as usize;
+    let mut connector = Connector::new()
+        .screen(screen_width, screen_height)
+        .credentials(".".to_string(), username.to_string(), password.to_string());
+    let client = match connector.connect(tcp) {
+        Ok(client) => client,
+        Err(e) => return format!("failed RDP connection to {}: {:?}", &addr, e).into(),
+    };
+
+    register_client(
+        client_ref,
+        Client {
+            rdp_client: client,
+            tcp_fd: tcp_fd,
+        },
+    );
+    CGOError::null()
+}
+
+#[repr(C)]
+pub struct CGOBitmap {
+    pub dest_left: u16,
+    pub dest_top: u16,
+    pub dest_right: u16,
+    pub dest_bottom: u16,
+    // Memory is freed on the Rust side.
+    pub data_ptr: *const u8,
+    pub data_len: usize,
+}
+
+impl TryFrom<BitmapEvent> for CGOBitmap {
+    type Error = Error;
+
+    fn try_from(e: BitmapEvent) -> Result<Self, Self::Error> {
+        let mut res = CGOBitmap {
+            dest_left: e.dest_left,
+            dest_top: e.dest_top,
+            dest_right: e.dest_right,
+            dest_bottom: e.dest_bottom,
+            data_ptr: std::ptr::null(),
+            data_len: 0,
+        };
+
+        // e.decompress consumes e, so we need to call it separately, after populating the fields
+        // above.
+        let data = if e.is_compress {
+            e.decompress()?
+        } else {
+            e.data
+        };
+        res.data_ptr = data.as_ptr();
+        res.data_len = data.len();
+
+        Ok(res)
+    }
+}
+
+// TODO: this is Linux-specific, also implement for Windows.
```

Comment:
You need to conditionally compile this for windows. Add this above the function below the comment.
`#[cfg(target_os = "linux")]`
Read more here https://doc.rust-lang.org/reference/conditional-compilation.html